File: leetcode/medium/249-group-shifted-strings/249-group-shifted-strings.py
from typing import List


class Solution:
    def getSequence(self, s: str) -> str:
        if not s:
            return ''

        if len(s) == 1:
            return '?'

        res = []

        for i in range(1, len(s)):
            diff = ord(s[i]) - ord(s[i - 1])

            if diff < 0:
                diff += 26

            res.append(str(diff))

        return ','.join(res)

    def groupStrings(self, strings: List[str]) -> List[List[str]]:
        dict = {}

        for s in strings:
            seq = self.getSequence(s)

            if seq not in dict:
                dict[seq] = []

            dict[seq].append(s)

        return list(dict.values())

File: leetcode/medium/249-group-shifted-strings/249-group-shifted-strings.ts
// 249-group-shifted-strings
// URL: https://leetcode.com/problems/group-shifted-strings/description/?envType=study-plan-v2&envId=premium-algo-100


function getSequenceFromString(str: string): string {
    const sequence = [];
  
    for (let i = 1; i < str.length; i++) {
      const diff = str.charCodeAt(i) - str.charCodeAt(i - 1);
      sequence.push(diff < 0 ? diff + 26 : diff);
    }
  
    return sequence.join(',');
  }
  
  function groupStrings(strings: string[]): string[][] {
    const hash: Map<string, string[]> = new Map();
  
    strings.forEach((str) => {
      const sequence: string = getSequenceFromString(str);
  
      if (hash.has(sequence)) {
        hash.get(sequence).push(str);
      } else {
        hash.set(sequence, [str]);
      }
  
    });
  
    return [...hash.values()];
  };
  